### ISSUE TITLE
C++ based modules in iot-edge

### ISSUE THREAD

**Issue by Senthamizhgp** [Azure/iot-edge#383]
Hi,

As per the current iot-edge framework, the modules implemented are ‘C’ language based modules like hello_world, simulated_device etc and I understand the main core iot-sdk-c is also based on C.

Considering few factors like modularity, maintenance etc the requirement is to develop modules in C++ language. Hence decided upon on creating C++ modules in iot-edge.  Yes agree compiler settings difference between C and C++ can cause changes in the modules (existing or newly created ones).  Ofcourse there can be two ways of achieving the C++ functionality.
-> C++ code : Implement the module itself completely in C++
-> Mix of C and C++ code : Implement the module structure in C (as in current iot-edge samples), have all other processing/validation etc separately using C++ and finally utilize them in C module (using wrapper interface for C++)
Which of the above do you suggest considering your iot-sdk framework (module creation, message passing etc) in C language? Hope that’s fine or do you see any problems? 

Kindly also provide the implications on usage of C++ based modules in Azure iot-edge if any. Also references if any.

In short basically can we implement the iot-edge modules/libraries in C++?

**Comment by Senthamizhgp** [comment]
Any suggestions/updates?

**Comment by avranju** [comment]
The edge run time requires that modules export the C function `Module_GetApi` that looks like this:

```
const MODULE_API* Module_GetApi(MODULE_API_VERSION gateway_api_version);
```

You can certainly implement your core module code in C++. You'll just need to make sure that you implement `Module_GetApi` under `extern C` (to turn off C++ name mangling) and the `MODULE_API` struct should be filled with function pointers to C functions. Beyond that you're free to implement everything else as you see fit.

**Comment by Senthamizhgp** [comment]
Ofcourse agree... Thanks for your response..
